src/control/server/storage_nvme.go

```diff
@@ -345,51 +330,136 @@ func (n *nvmeStorage) Format(i int, resp *pb.FormatStorageResp) {
+// Update attempts to update firmware on NVMe controllers attached to a
+// given server identified by PCI addresses as specified in config file.
+// Update populates resp NvmeControllerResult for each NVMe controller
+// specified in config file bdev_list param.
+//
+// Firmware will only be updated if the controller the current fw rev
+// and model match the "startRev" and "model" fn parameters respectively.
+// Path and slot params refer to the fw image file location and controller
+// firmware register to update respectively.
 //
-// Firmware will only be updated if the controller reports the current fw rev
-// to be the same as the "startRev" fn parameter. path and slot refer to the
-// fw image file location and controller fw register to update respectively.
+// One result with empty Pciaddr will be reported if there are preliminary
+// errors occurring before devices could be accessed. Otherwise a result will
+// be populated for each device in bdev_list.
+//
+// TODO: take model name/number to verify we are updating expected model
 func (n *nvmeStorage) Update(
-	i int, startRev string, path string, slot int32) error {
+	i int, req *pb.UpdateNvmeParams, results *([]*pb.NvmeControllerResult)) {
 
+	var pciAddr string
 	srv := n.config.Servers[i]
+	log.Debugf("performing firmware update on NVMe controllers")
+
+	// appends results to response to provide update specific function
+	addCretUpdate := func(status pb.ResponseStatus, errMsg string) {
+		// log depth should be stack layer registering result
+		*results = append(
+			*results,
+			newCret(
+				"update", pciAddr, status, errMsg,
+				common.UtilLogDepth+1))
+	}
 
 	if !n.initialized {
-		return errors.New(msgBdevNotInited)
+		addCretUpdate(
+			pb.ResponseStatus_CTRL_ERR_APP, msgBdevNotInited)
+		return
 	}
 
 	switch srv.BdevClass {
 	case bdNVMe:
-		for _, pciAddr := range srv.BdevList {
+		for _, pciAddr = range srv.BdevList {
 			if pciAddr == "" {
+				addCretUpdate(
+					pb.ResponseStatus_CTRL_ERR_CONF,
+					msgBdevEmpty)
 				continue
 			}
 
 			ctrlr := n.getController(pciAddr)
 			if ctrlr == nil {
+				addCretUpdate(
+					pb.ResponseStatus_CTRL_ERR_NVME,
+					pciAddr+": "+msgBdevNotFound)
 				continue
 			}
-			if ctrlr.Fwrev != startRev {
+
+			if ctrlr.Model != req.Model {
+				addCretUpdate(
+					pb.ResponseStatus_CTRL_ERR_NVME,
+					fmt.Sprintf(
+						pciAddr+": "+
+							msgBdevModelMismatch+
+							" want %s, have %s",
+						req.Model, ctrlr.Model))
 				continue
 			}
 
-			cs, ns, err := n.nvme.Update(pciAddr, path, slot)
+			if ctrlr.Fwrev != req.Startrev {
+				addCretUpdate(
+					pb.ResponseStatus_CTRL_ERR_NVME,
+					fmt.Sprintf(
+						pciAddr+": "+
+							msgBdevFwrevStartMismatch+
+							" want %s, have %s",
+						req.Startrev, ctrlr.Fwrev))
+				continue
+			}
+
+			log.Debugf(
+				"updating firmware (%s->%s) on nvme "+
+					"controller at %s, may take several minutes!",
+				req.Startrev, ctrlr.Fwrev, pciAddr)
```

Comment:
The message implies that we will see a change from firmware version A to B, but req.Startrev and ctrlr.Fwrev are the same value here.  Do you have access to the expected firmware value after the upgrade?  If so, then let's print that instead, or change the message so it doesn't reference the firmware versions.

Reply:
good catch